sklearn/preprocessing/_csr_polynomial_expansion.pyx

```diff
@@ -118,9 +128,18 @@ def _csr_polynomial_expansion(ndarray[DATA_T, ndim=1] data,
     cdef ndarray[INDEX_T, ndim=1] expanded_indptr = ndarray(
         shape=num_rows + 1, dtype=indptr.dtype)
 
+    # if the expanded_dimensionality is too big, we need to cast
+    # indices to int64. If total_nnz is too big, we need to cast
+    # expanded_indptr to int64 too.
+    # TODO: check that the casting is indeed done
+    if expanded_dimensionality > MAX_INT32:
+        indices = np.array(indices, dtype=np.int64)
+        expanded_indices = np.array(expanded_indices, dtype=np.int64)
+    if total_nnz > MAX_INT32:
+        expanded_indptr = np.array(expanded_indptr, dtype=np.int64)
```

Comment:
This could lead to a csr with indicies and indptr not having the same dtype. I do not know if scipy supports that.